### DISCUSSION TITLE
Harriss Spiral Design Turtle Graphics

### DISCUSSION THREAD

**Discussion by vsquared** [py5coding/py5generator#545] [Show and tell]
This demo creates a Harriss Spiral design using Turtle Graphics for Processing with recursion:
```python
# Reference for Turtle Graphics in Processing
# https://gist.githubusercontent.com/nataliefreed/8483050/raw/9e3f1d0f44bcb0c872762e4b984358d375e7b5fa/turtle.pde
# Use imported mode for py5

end = 0.0
loc = 0.0
newLoc = 0.0
h = 600.0
w = (h/1.325)/1.325
orientation = radians(90)
rotation = -45
showLines = False

def harriss(iteration):
    global w
    
    if(iteration > 0):
        iteration -= 1
        fd(w,iteration)
        lt(radians(90))
        w = w/1.325
        harriss(iteration)
        
def setup():
    global loc
    global showLines
    
    size(800, 750)
    surface = get_surface()
    surface.set_title("Harriss Spiral Design")
    showLines = False
    loc = Py5Vector(width/2 + 100,height/2 + 200)
    no_loop()
    
def draw():
    global orientation
    global rotation
    global h
    global w

# Base Spiral
    harriss(7)
# Right Upper Spiral
    jumpTo(495, 230)
    h = 300.0
    w = (h/1.325)/1.325
    orientation = radians(0)
    rotation = 45
    harriss(8)
# Left Upper Spiral
    jumpTo(240, 238)
    h = 250
    w = (h/1.325)/1.325
    rotation = -45
    orientation = radians(90)
    harriss(7)
# Left Lower Spiral
    jumpTo(240, 425)
    h = 200
    w = (h/1.325)/1.325
    rotation = 225
    orientation = radians(180)
    harriss(6);
# Center Lower Spiral
    jumpTo(390, 425)
    h = 175
    w = (h/1.325)/1.325
    rotation = 135
    orientation = radians(-90)
    harriss(5)    
    
# ======= Turtle Graphics Functions =========
    
def fd(w,iteration):
    global loc
    global newLoc
    global end
    global rotation
    
    polar(w,orientation)
    end = loc + newLoc

    if(showLines):
        stroke(0)
        stroke_weight(1)
        line(loc.x,loc.y,end.x,end.y)
    
    no_fill()
    stroke(random(255),0,random(255))
    stroke_weight(12);
    if ((iteration == 7) or (iteration == 3)):
        arc(loc.x + w/2, loc.y - w/2, w*1.414, w*1.414, radians(rotation), radians(rotation + 90));
    if ((iteration == 6) or (iteration == 2)):
        arc(loc.x - w/2, loc.y - w/2, w*1.414, w*1.414, radians(rotation), radians(rotation + 90));  
    if ((iteration == 5) or (iteration == 1)):
        arc(loc.x - w/2, loc.y + w/2, w*1.414, w*1.414, radians(rotation), radians(rotation + 90));  
    if ((iteration == 4) or (iteration == 0)):
        arc(loc.x + w/2, loc.y + w/2, w*1.414, w*1.414, radians(rotation), radians(rotation + 90));
          
    loc = end
    rotation -= 90
        
def lt(theta):
    global orientation
    orientation += theta

def jumpTo(x,y):
    global loc
    loc = Py5Vector(x,y)
    
def polar(r,theta):
    global newLoc
    newLoc = Py5Vector(r*cos(theta),r*sin(-theta))

```
**Output:**

<img width="726" alt="py5" src="https://github.com/user-attachments/assets/0015527d-75ec-4382-bb42-0f3ba821f3f4">

**Comment by hx2A** [comment]
I just opened a new issue #554 . I believe your call to `no_loop()` should prevent `draw()` from being called. That's what happens for the equivalent code in module mode. You've inadvertently discovered a bug.

Anyhow, I thought this was interesting and converted it because I wanted to play around with it using the new Live Coding feature. Live Coding (currently) only supports module mode. It is fun to play around with this using Live Coding and watch how the lines move around.

In the future I hope to get Live Coding working for imported mode also.

**Reply by vsquared** [reply]
All those modes are a mystery to me.  I just know it works when I use 'Imported mode for py5'.  No 'main', no 'runSketch', no prefixes,  just simple code that works.   As far as the bug goes, if you change the function draw() to drawSpiral() and replace no_loop() with a call to drawSpiral() it still works.  I can edit the post if it will avoid confusion in the future.  Draw() is called once in Processing if you also use noLoop() in setup() which is how I wrote this demo initially, so I was just using what I was used to.  Did not know py5 was supposed to work differently.  Sometimes it's good when you don't know how things are supposed to work.

**Reply by hx2A** [reply]
> just simple code that works

Music to my ears. This is what Python is all about, and what I aspire for py5 to be.

>  Draw() is called once in Processing if you also use noLoop() in setup() which is how I wrote this demo initially, so I was just using what I was used to. Did not know py5 was supposed to work differently.

No, you were correct here! Processing and py5 both will call `draw()` once when you use `no_loop()` in `setup()`. The problem is in Live Coding. Live Coding has to do some extra stuff so that `no_loop()` behaves correctly, and I misunderstood how it was supposed to work. It actually was me who didn't realize that it worked differently, so I coded the Live Coding functionality incorrectly. This thread brought my mistake to my attention.

> Sometimes it's good when you don't know how things are supposed to work.

Yes, indeed. Sometimes not knowing stuff contributes to creativity and new ideas.

Also, I learn a lot from the people who use py5, such as yourself. I learned something new about Processing today. Thanks so much for taking the time to share your Harriss Spiral examples.
